### ISSUE TITLE
Squiz.Commenting.FunctionComment.InvalidTypeHint on Scalar types on PHP7

### ISSUE THREAD

**Issue by rbairwell** [squizlabs/PHP_CodeSniffer#805]
As you probably know, PHP 7 introduced scalar type declarations (previously known as TypeHints) so that "string", "int", "float", "bool" can be declared in a method parameter and return type.

PHPCS does cope with these when they have a leading backslash such as:
````php
namespace x\y\z;

class test {
/**
 * @param \string $name
 */
 public function (\string $name)
....
````

However, a recent commit relating to PHP 7.0.1 (see https://github.com/php/php-src/commit/569763cb1ac67f56e7743062ca8b3b7c650c1254 "Fixed \int (or generally every scalar type name with leading backslash) to not be accepted as type name. (Bob)" and https://github.com/php/php-src/commit/00865ae22f2c5fdee9e500ce79d442467e0a0899) which means the backslashed scalars/builtins are technically incorrect in PHP 7.0.1 onwards (and throws a fatal error if set).

Running PHPCS without the variables backslashes results in a load of:
````
Unknown type hint "string" found for $keyName
 (Squiz.Commenting.FunctionComment.InvalidTypeHint)
````

Can the code sniffer allow these types without the backslash?

[tested using PHPCS commit 43ac634 , PHP 7.0.1 on an upgraded rasmus PHPdev VM]

**Comment by gsherwood** [comment]
I've fixed that up now. Thanks for reporting it.

Error will still be thrown if you are using PHP < 7 though.

**Comment by rbairwell** [comment]
Thank you very much for the prompt fix... However, I'm still getting (under PHP 7.0.1 using latest fe8c328 commit from dev-master)
```
  52 | ERROR   | Expected "boolean" but found "bool" for @var tag in
     |         | member variable comment
     |         | (Squiz.Commenting.VariableComment.IncorrectVarType)
````
for code such as:
````php
/**
 * Something something
*
* @var bool
*/
public $xyz=true;
````

Looking at the change committed, I can't see why this is occurring (as boolean is being mapped to bool)

**Comment by aik099** [comment]
This isn't what bug report is about. Existing sniff require full type name in `@var` command (`integer`, not `int` and `boolean` not `bool`).

**Comment by gsherwood** [comment]
> Existing sniff require full type name in @var command (integer, not int and boolean not bool). This is how it was implemented, not a bug.

What @aik099 said. The Squiz sniff wants you to use full type names in the function comments. I've left that require intact while fixing this issue.

**Comment by giginos** [comment]
```
/**
* @param int $id
* @return FooClass
*/
public function setId(int $id): FooClass
{
	$this->id = $id;
	return $this;
}
```

-> Will cause a "Unknown type" hint.


```
/**
* @param integer $id
* @return FooClass
*/
public function setId(int $id): FooClass
{
	$this->id = $id;
	return $this;
}
```

-> Will accept the scalar type but phpcs does not like "@param integer" at all.

So it would be great if the first sample would be evaluated as valid.

**Comment by gsherwood** [comment]
> -> Will cause a "Unknown type" hint.

It doesn't cause an unknown type hint error on either PHP 5 or 7. I've tested PHPCS 2.6.2 and master.

**Comment by giginos** [comment]
Thx. You're right @gsherwood t. It depends on our codeing standard. phpcs + PSR2 does the job.

**Comment by FGM** [comment]
I still have this error on phpcs 2.7.1 run under PHP 7.0.14, although as I understand from this issue, it should no longer appear:

 20 | ERROR | [ ] Unknown type hint "string" found for $directory

The code is like:
```php
  /**
   * RequirementsLoader constructor.
   *
   * @param string $directory
   *   The directory from which to load the composer.json file.
   */
  public function __construct(string $directory) {}
```

**Comment by gsherwood** [comment]
@FGM None of the included standards throw this error, so it is possibly an issue with the Drupal sniffs. Run PHPCS with the `-s` command line option and it will tell you what sniff caused the error. If it is a Drupal one, report that to whoever maintains the Drupal standard. If it is a sniff from one of the included standards, let me know and I'll take a look.

**Comment by FGM** [comment]
@gsherwood indeed; the result was "Drupal.Commenting.FunctionComment.InvalidTypeHint". Thanks for your help.

**Comment by guix77** [comment]
`@param \int $uid`
